### PR TITLE
[Net] Retry feeler connection if OpenNetworkConnection fails in under 1ms.

### PR DESCRIPTION
**Pull request by pstratem**
Avoids issues with feelers to ip space we don't know how to route.

### CODE REVIEW

**Comment by laanwj** [comment]
Doesn't this need a protection to only do this once?

**Comment by pstratem** [comment]
@laanwj The address being connected it selected new each time the loop runs, because of how the logic works it won't try the same thing in a loop.

**Comment by laanwj** [comment]
Sure, so it won't 'DoS' a single host, but what if there is a connection problem with a whole IP range? E.g. from the local network to the internet. Then you don't want to keep retrying every ms either.

**Comment by pstratem** [comment]
@laanwj It re-tries every 500ms

**Comment by gmaxwell** [comment]
ACK.  Maybe it could use a comment to point out that it uses and depends on the outer loop sleep.

**Comment by pstratem** [comment]
rebased and comment added

File: src/net.cpp [hunk 1]
```diff
@@ -1734,7 +1734,15 @@ void ThreadOpenConnections()
...
 
-            OpenNetworkConnection(addrConnect, (int)setConnected.size() >= std::min(nMaxConnections - 1, 2), &grant, NULL, false, fFeeler);
+            int64_t nStartTime = GetTimeMicros();
+
+            bool fRet = OpenNetworkConnection(addrConnect, (int)setConnected.size() >= std::min(nMaxConnections - 1, 2), &grant, NULL, false, fFeeler);
+
+            // Retry feeler connection if OpenNetworkConnection fails in under 1ms.
+            // Avoids issues with feelers to IP space we don't know how to route.
+            if (!fRet && GetTimeMicros() - nStartTime < 1000) {
+                nNextFeeler = GetTime();
```
Comment (sipa) [inline src/net.cpp]: nNextFeeler is in microseconds, so you should reuse the value of the call to GetTimeMicros instead (or just assign 0, which will have the same effect).

**Comment by gmaxwell** [comment]
ping @pstratem

File: src/net.cpp [hunk 2]
```diff
@@ -1734,7 +1734,15 @@ void ThreadOpenConnections()
                 LogPrint("net", "Making feeler connection to %s\n", addrConnect.ToString());
             }
 
-            OpenNetworkConnection(addrConnect, (int)setConnected.size() >= std::min(nMaxConnections - 1, 2), &grant, NULL, false, fFeeler);
+            int64_t nStartTime = GetTimeMicros();
+
+            bool fRet = OpenNetworkConnection(addrConnect, (int)setConnected.size() >= std::min(nMaxConnections - 1, 2), &grant, NULL, false, fFeeler);
```
Comment (rebroad) [inline src/net.cpp]: nit: "ret" seems to be the standard so far, with 69 instances, compared to only 10 instances of "fRet" in this project.

Comment (pstratem) [inline src/net.cpp]: I'm matching the style of code around this, all variable names start with their type.

File: src/net.cpp (hunk 1 above)
Comment (pstratem) [inline src/net.cpp]: Fixed.

**Comment by rebroad** [comment]
tACK

File: src/net.cpp (hunk 1 above)
Comment (sipa) [inline src/net.cpp]: Not fixed.

Comment (pstratem) [inline src/net.cpp]: I changed it to be 0 but i've now changed it to GetTimeMicros()

Comment (TheBlueMatt) [inline src/net.cpp]: Did you forget to push?

File: src/net.cpp [hunk 3]
```diff
@@ -1701,7 +1701,16 @@ void CConnman::ThreadOpenConnections()
...
+
+            // Retry feeler connection if OpenNetworkConnection fails in under 1ms.
+            // Avoids issues with feelers to IP space we don't know how to route.
+            // The outer loop has a 500ms delay, we rely on that delay here
+            if (!fRet && GetTimeMicros() - nStartTime < 1000) {
```
Comment (TheBlueMatt) [inline src/net.cpp]: Does this need an && fFeeler in ti?

**Review by TheBlueMatt — COMMENTED** [review]
Can you provide more justification for this change? Why do only feelers matter (or is it all connections?)? What are the "issues" you are referring to?

**Comment by pstratem** [comment]
@TheBlueMatt Currently if the address selected for the feeler logic is in ipv6 space and you can only route ipv4 you will wait about 120 seconds to retry, this is simply to avoid that delay.

**Comment by sipa** [comment]
@pstratem Can you address https://github.com/bitcoin/bitcoin/pull/8695/files#r99852785 ?

**Comment by sipa** [comment]
@pstratem Is this going anywhere?

**Comment by jnewbery** [comment]
No activity. Closing with 'up for grabs' for now. @pstratem - feel free to re-open if you pick this up again.

**Pull request closed without merging**